### DISCUSSION TITLE
Choice of EnsureSuccess() / Exceptions

### DISCUSSION THREAD

**Discussion by srs-adamr** [nats-io/nats.net#492] [General]
I wanted to start a discussion regarding the choice and use of Exceptions for the results of a number of methods in the SDK. 

For example:
```
var ack = await js.PublishAsync("orders.new.1", order, serializer: orderSerializer);

ack.EnsureSuccess();
```
I would like to understand the design choice here to provide a `bool` for Duplicate but not one for `Success` and instead rely on throwing of an Exception?

**Comment by mtmk** [comment]
`PubAckResponse` is essentially the object returned by the server so we don't add anything to is. `EnsureSuccess()` is and extension method for convenience to make sure message is saved before carrying on. This should suit most scenarios. When msg-id is used then you have to take care of duplicates based on your application requirements which would also mean you'd need to check for errors as well e.g. `msg.Error != null`.

What is your use case? if we also provide something like `IsSuccess()` extension would that work for you?

**Reply by srs-adamr** [reply]
> When msg-id is used then you have to take care of duplicates based on your application requirements which would also mean you'd need to check for errors as well e.g. `msg.Error != null`.

Yes, this is what I was thinking about. Avoiding the use of Exceptions for performance purposes. For example, taking the `HttpClient` as an example:

```
var result = await httpClient.PostAsJsonAsync("123.com", obj);
if (result.IsSuccessStatusCode)
{

}
```

> What is your use case? if we also provide something like `IsSuccess()` extension would that work for you?

This is exactly what I was thinking about 👍

**Reply by mtmk** [reply]
Just created a PR for that https://github.com/nats-io/nats.net.v2/pull/493

@srs-adamr thanks for pointing this out 💯
